lib/matplotlib/widgets.py

```diff
@@ -2175,32 +2195,60 @@ def __init__(self, ax, onselect=None, useblit=True, lineprops=None,
         self.line.set_visible(False)
         self.ax.add_line(self.line)
         self.artists = [self.line]
+        self.persist = persist
 
     def onpress(self, event):
         self.press(event)
 
     def _press(self, event):
-        self.verts = [self._get_data(event)]
+        if (not event.key == self.state_modifier_keys['square'] or
+                self.verts is None):
+            self.verts = [(event.xdata, event.ydata)]
+
         self.line.set_visible(True)
 
     def onrelease(self, event):
         self.release(event)
 
-    def _release(self, event):
-        if self.verts is not None:
-            self.verts.append(self._get_data(event))
-            self.onselect(self.verts)
+    def _clear(self, event):
+        self.verts = None
         self.line.set_data([[], []])
         self.line.set_visible(False)
+        self.update()
+
+    def _finish(self, event):
+        self.verts.append(self.verts[0])
+        self.line.set_data(list(zip(*self.verts)))
+        self.update()
+        self.onselect(self.verts)
+        if not self.persist:
+            self._clear(event)
         self.verts = None
 
-    def _onmove(self, event):
+    def _release(self, event):
         if self.verts is None:
             return
-        self.verts.append(self._get_data(event))
+        self.verts.append((event.xdata, event.ydata))
+        if event.key != self.state_modifier_keys['square']:
```

Comment:
so, then, shouldn't it be "polygon" here and further down, too? Where does input normalization happen to guarantee the presence the "polygon" entry?